client.lua

```diff
@@ -0,0 +1,40 @@
+local config = require 'config.client'
+
+---@type table<string, integer>
+local states = {}
+
+---Sets a boolean state. Runs its onSet function if the state is not previously set and the onSet function exists in config
+---@param name string
+---@return boolean firstSetter true if the state had not previously been set
+local function setState(name)
+    if not states[name] then
+        if config.states[name]?.onSet then
+            config.states[name].onSet()
+        end
+        states[name] = 1
+        return true
+    else
+        states[name] += 1
+        return false
+    end
+end
+
+---Releases a boolean state. Runs its onRelease function if this is the last call to release the state and the release function exists in config.
+---@param name string
+---@return boolean lastReleaser true if the state is fully released after this function
+local function releaseState(name)
+    if not states[name] then states[name] = 0 end
```

Comment:
This would only happen in the case of a misbehaving caller; one that calls release more than they call SetState. In that situation, if we call onRelease, then the behavior is the same as it is today; the state is released sooner than it should be. If we don't call onRelease, we're pushing this situation to the callers to sort out, which they would probably do by releasing the state anyway.

The other option is to throw an error in this situation.

Reply:
On second thought, made this case a no-op and added a force parameter as there are cases where we definitely want to release the state entirely even if other resources have a lock.